Fraud Detection

Explanatory data analysis

In [15]:
import pandas as pd
import sidetable as stb
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import ipaddress 


In [26]:
fraud_df = pd.read_csv('../data/Fraud_Data.csv')
ip_df = pd.read_csv('../data/IpAddress_to_Country.csv')
creditcard_df = pd.read_csv('../data/creditcard.csv')

print("Fraud_Data shape:", fraud_df.shape)
print("IP Data shape:", ip_df.shape)
print("creditcard:", creditcard_df.shape)


Fraud_Data shape: (151112, 11)
IP Data shape: (138846, 3)
creditcard: (284807, 31)


In [11]:
fraud_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 151112 entries, 0 to 151111
Data columns (total 11 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   user_id         151112 non-null  int64  
 1   signup_time     151112 non-null  object 
 2   purchase_time   151112 non-null  object 
 3   purchase_value  151112 non-null  int64  
 4   device_id       151112 non-null  object 
 5   source          151112 non-null  object 
 6   browser         151112 non-null  object 
 7   sex             151112 non-null  object 
 8   age             151112 non-null  int64  
 9   ip_address      151112 non-null  float64
 10  class           151112 non-null  int64  
dtypes: float64(1), int64(4), object(6)
memory usage: 12.7+ MB


In [12]:
fraud_df.describe()

user_id  purchase_value            age    ip_address  \
count  151112.000000   151112.000000  151112.000000  1.511120e+05   
mean   200171.040970       36.935372      33.140704  2.152145e+09   
std    115369.285024       18.322762       8.617733  1.248497e+09   
min         2.000000        9.000000      18.000000  5.209350e+04   
25%    100642.500000       22.000000      27.000000  1.085934e+09   
50%    199958.000000       35.000000      33.000000  2.154770e+09   
75%    300054.000000       49.000000      39.000000  3.243258e+09   
max    400000.000000      154.000000      76.000000  4.294850e+09   

               class  
count  151112.000000  
mean        0.093646  
std         0.291336  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000

In [16]:
fraud_df.stb.counts()

count  unique            most_freq  most_freq_count  \
sex             151112       2                    M            88293   
class           151112       2                    0           136961   
source          151112       3                  SEO            60615   
browser         151112       5               Chrome            61432   
age             151112      58                   32             6666   
purchase_value  151112     122                   28             3328   
device_id       151112  137956        CQTUVBYIWWWBC               20   
ip_address      151112  143512     1502818419.73176               20   
purchase_time   151112  150679  2015-07-17 23:22:55                3   
user_id         151112  151112                22058                1   
signup_time     151112  151112  2015-02-24 22:55:49                1   

                         least_freq  least_freq_count  
sex                               F             62819  
class                             1             14151  
source                       Direct             30616  
browser                       Opera              3676  
age                              73                 1  
purchase_value                  123                 1  
device_id             GCYGGWEYXYVEW                 1  
ip_address          732758368.79972                 1  
purchase_time   2015-08-30 08:09:07                 1  
user_id                       22058                 1  
signup_time     2015-02-24 22:55:49                 1

In [17]:
ip_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 138846 entries, 0 to 138845
Data columns (total 3 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   lower_bound_ip_address  138846 non-null  float64
 1   upper_bound_ip_address  138846 non-null  int64  
 2   country                 138846 non-null  object 
dtypes: float64(1), int64(1), object(1)
memory usage: 3.2+ MB


In [34]:
ip_df.describe()
ip_df.shape

(138846, 3)

In [19]:
ip_df.stb.counts()

count  unique      most_freq  most_freq_count  \
country                 138846     235  United States            46868   
lower_bound_ip_address  138846  138846     16777216.0                1   
upper_bound_ip_address  138846  138846       16777471                1   

                        least_freq  least_freq_count  
country                    Tokelau                 1  
lower_bound_ip_address  16777216.0                 1  
upper_bound_ip_address    16777471                 1

In [20]:
creditcard_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [38]:
creditcard_df.describe()

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  1.175161e-15  3.384974e-16 -1.379537e-15  2.094852e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   1.021879e-15  1.494498e-15 -5.620335e-16  1.149614e-16 -2.414189e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.628620e-16 -3.576577e-16  2.618565e-16  4.473914e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   5.109395e-16  1.686100e-15 -3.661401e-16 -1.227452e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

In [22]:
creditcard_df[['Class' , 'Amount' , 'Time','V1']].stb.counts()

count  unique      most_freq  most_freq_count     least_freq  \
Class   284807       2       0.000000           284315       1.000000   
Amount  284807   32767       1.000000            13688     534.600000   
Time    284807  124592  163152.000000               36  172792.000000   
V1      284807  275663       1.245674               77      -1.015042   

        least_freq_count  
Class                492  
Amount                 1  
Time                   1  
V1                     1

In [25]:
fraud_df.isna().sum()

user_id           0
signup_time       0
purchase_time     0
purchase_value    0
device_id         0
source            0
browser           0
sex               0
age               0
ip_address        0
class             0
dtype: int64

In [46]:
fraud_df.duplicated().sum()

np.int64(0)

In [42]:
fraud_df.stb.missing()

missing   total  percent
user_id               0  151112      0.0
signup_time           0  151112      0.0
purchase_time         0  151112      0.0
purchase_value        0  151112      0.0
device_id             0  151112      0.0
source                0  151112      0.0
browser               0  151112      0.0
sex                   0  151112      0.0
age                   0  151112      0.0
ip_address            0  151112      0.0
class                 0  151112      0.0

Dataset Overview
---------------------
- Total rows: 151112
- Total columns: 11

No Duplicated data found in the dataset.

No missing data found in the dataset.

In [31]:
ip_df.isna().sum()


lower_bound_ip_address    0
upper_bound_ip_address    0
country                   0
dtype: int64

In [45]:
ip_df.duplicated().sum()


np.int64(0)

In [41]:
ip_df.stb.missing()

missing   total  percent
lower_bound_ip_address        0  138846      0.0
upper_bound_ip_address        0  138846      0.0
country                       0  138846      0.0

Dataset Overview
---------------------
- Total rows: 138846
- Total columns: 3

No Duplicated data found in the dataset.

No missing data found in the dataset.

In [35]:
creditcard_df.isna().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [44]:
creditcard_df.duplicated().sum()

np.int64(1081)

In [43]:
creditcard_df.stb.missing()

missing   total  percent
Time          0  284807      0.0
V1            0  284807      0.0
V2            0  284807      0.0
V3            0  284807      0.0
V4            0  284807      0.0
V5            0  284807      0.0
V6            0  284807      0.0
V7            0  284807      0.0
V8            0  284807      0.0
V9            0  284807      0.0
V10           0  284807      0.0
V11           0  284807      0.0
V12           0  284807      0.0
V13           0  284807      0.0
V14           0  284807      0.0
V15           0  284807      0.0
V16           0  284807      0.0
V17           0  284807      0.0
V18           0  284807      0.0
V19           0  284807      0.0
V20           0  284807      0.0
V21           0  284807      0.0
V22           0  284807      0.0
V23           0  284807      0.0
V24           0  284807      0.0
V25           0  284807      0.0
V26           0  284807      0.0
V27           0  284807      0.0
V28           0  284807      0.0
Amount        0  284807      0.0
Class         0  284807      0.0

Dataset Overview
---------------------
- Total rows: 284807
- Total columns: 31

there is  Duplicated data found in the dataset.(1081)

No missing data found in the dataset.

In [47]:
creditcard_df.drop_duplicates(inplace=True)

Data Type correction

In [50]:
fraud_df['signup_time'] = pd.to_datetime(fraud_df['signup_time'])
fraud_df['purchase_time'] = pd.to_datetime(fraud_df['purchase_time'])
fraud_df['ip_address'] = fraud_df['ip_address'].astype('int64')

In [52]:
ip_df['lower_bound_ip_address'] = ip_df['lower_bound_ip_address'].astype('int64')